modules/e2ee/Worker.js

```diff
@@ -149,16 +162,30 @@ class Context {
      * @param {CryptoKey} key
      * @param {Number} keyIndex
      */
-    async setKey(key, keyIndex) {
+    async setKey(keyBytes, keyIndex) {
         this._currentKeyIndex = keyIndex % this._cryptoKeyRing.length;
-        if (key) {
-            this._cryptoKeyRing[this._currentKeyIndex] = await deriveKeys(key);
+        if (keyBytes) {
+            // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/importKey
+            const material = await crypto.subtle.importKey('raw', keyBytes,
+                'HKDF', false, [ 'deriveBits', 'deriveKey' ]);
+
+            this._cryptoKeyRing[this._currentKeyIndex] = await deriveKeys(material);
         } else {
             this._cryptoKeyRing[this._currentKeyIndex] = false;
         }
         this._sendCount = 0n; // Reset the send count (bigint).
     }
 
+    /**
+     * Ratchets a key forward one step.
+     */
+    async ratchet() {
```

Comment:
If we get the ratchet key early, when we derive all keys, this would just be:

```
const keys = this._cryptoKeyRing[this._currentKeyIndex];

this.setKey(keys.ratchetMaterial, this._currentKeyIndex);
```

The benefits of this may become more obvious if we factor all key shenanigans to some helper class.